src/Datadog.Trace/Tracer.cs

```diff
@@ -682,7 +714,17 @@ private void CurrentDomain_UnhandledException(object sender, UnhandledExceptionE
 
         private void Console_CancelKeyPress(object sender, ConsoleCancelEventArgs e)
         {
-            RunShutdownTasks();
+            // Console Cancel KeyPress is raised on SIGINT and SIGQUIT signal,
+            // here we ensure we flush and not close the writer.
+            // There's a possibility of a graceful shutdown handled by ProcessExit.
+            try
+            {
+                _agentWriter.FlushTracesAsync().GetAwaiter().GetResult();
```

Comment:
Is there any danger of a deadlock here? For example, if there's a custom `SynchronizationContext`?

Reply:
Adding the synchronization context check causes no harm. In the other hand I’ll be suprised if this method could be called with a synchronization context at all. Being raised by an OS notification.

Reply:
Yeah, it doesn't seems to use any syncronization context, the signals comes directly from the OS and in .NET Framework is handled by queueing to the [ThreadPool](https://referencesource.microsoft.com/#mscorlib/system/console.cs,1758), in .NET 5 is handled from a new [background thread (because the actual one is called from unmanaged code)](https://source.dot.net/#System.Console/System/ConsolePal.Unix.cs,1469)

In .NET Core 2.1: [Unix](https://github.com/dotnet/corefx/blob/release/2.1/src/System.Console/src/System/ConsolePal.Unix.cs#L1253) / Windows the delegate is registered by [this interop](https://github.com/dotnet/corefx/blob/release/2.1/src/Common/src/Interop/Windows/kernel32/Interop.SetConsoleCtrlHandler.cs#L19)

I don't think adding the `SyncronizationContext` check will change anything here.

Reply:
Because causes no harm, done in d39e660